### PR TITLE
Ellips reader and eln_data

### PR DESCRIPTION
**Pull request by cmmngr**
Updated ellips reader and metadata file (eln_data.yaml)

### CODE REVIEW

File: pynxtools/dataconverter/readers/ellips/mock.py [hunk 1]
```diff
@@ -47,10 +47,10 @@ class MockEllips():
 
     def __init__(self, data_template) -> None:
         self.data = data_template["measured_data"]
-        self.wavelength = data_template["spectrometer/wavelength"]
+        self.wavelength = data_template["data_collection/NAME_spectrum[wavelength_spectrum]"]
```
Comment (sanbrock) [inline pynxtools/dataconverter/readers/ellips/mock.py:50]: Is `data_collection` used in the appdef or it should rather be something like `DATA[data_collection]` here?

Comment (cmmngr) [reply pynxtools/dataconverter/readers/ellips/mock.py:50]: `data_collection` is used in the app def. It's `data_collection(NXprocess)`.

File: pynxtools/dataconverter/readers/ellips/reader.py [hunk 2]
```diff
@@ -24,10 +24,78 @@
...
+    'parameters',
+    'plot_name',
+    'sep',
+    'skip',
+    'spectrum_type',
+    'spectrum_unit'
+]
+
+REPLACE_NESTED = {
+    'INSTRUMENT[instrument]': 'INSTRUMENT[instrument]',
```
Comment (domna) [inline pynxtools/dataconverter/readers/ellips/reader.py:95]: This seems to do nothing here.

File: pynxtools/dataconverter/readers/ellips/reader.py [hunk 3]
```diff
@@ -192,6 +283,87 @@ def data_set_dims(whole_data):
...
+        index += counts[0]
+        block_idx.append(index)
+
+    # derived parameters:
+    # takes last but one column from the right (skips empty columns):
+    data_index = 1
+    temp = whole_data[header["colnames"][-data_index]].to_numpy()[
+        block_idx[-1] - 1].astype("float64")
+
+    while temp * 0 != 0:
```
Comment (domna) [inline pynxtools/dataconverter/readers/ellips/reader.py:308]: This is always False.

File: pynxtools/dataconverter/readers/ellips/reader.py [hunk 4]
```diff
@@ -217,6 +389,10 @@ def populate_header_dict_with_datasets(file_paths, is_mock=False):
         """
         header, data_file = populate_header_dict(file_paths)
 
+        # read first N lines of the data file:
+        # with open(data_file) as input_file:
```
Comment (domna) [inline pynxtools/dataconverter/readers/ellips/reader.py:393]: Commented code should just be removed.

File: pynxtools/dataconverter/readers/ellips/reader.py [hunk 5]
```diff
@@ -225,63 +401,33 @@ def populate_header_dict_with_datasets(file_paths, is_mock=False):
...
-                    val.append(f"{key}_{int(angle[1])}deg")
-                index += counts[angle[0]]
-                block_idx.append(index)
 
+        if "atom_types" not in header:
+            header["atom_types"] = extract_atom_types(header["chemical_formula"])
         # Atom types: Convert str to list if atom_types is not a list:
-        if isinstance(header["atom_types"], str):
-            header["atom_types"] = header["atom_types"].split(",")
+        # if isinstance(header["atom_types"], str):
```
Comment (domna) [inline pynxtools/dataconverter/readers/ellips/reader.py:427]: Commented code should just be removed

File: pynxtools/dataconverter/readers/utils.py [hunk 6]
```diff
@@ -41,6 +41,7 @@ class FlattenSettings():
     parent_key: str = "/ENTRY[entry]"
     sep: str = "/"
     is_in_section: bool = False
+    black_list: Optional[list] = None
```
Comment (domna) [inline pynxtools/dataconverter/readers/utils.py:44]: I'm thinking that we may should use a better name. Something like `ignore_keys`. What do you think?

Comment (cmmngr) [reply pynxtools/dataconverter/readers/utils.py:44]: agreed

File: pynxtools/dataconverter/readers/ellips/reader.py (hunk 2 above)
Comment (cmmngr) [reply pynxtools/dataconverter/readers/ellips/reader.py:95]: Right. This is a placeholder (for an empty list I got an error message...). After updating the ELN schema, some keys to be replaced will be added here.

File: pynxtools/dataconverter/readers/ellips/reader.py (hunk 3 above)
Comment (cmmngr) [reply pynxtools/dataconverter/readers/ellips/reader.py:308]: not if `temp = nan`. There's probably a better way to write that. Suggestions?

Comment (domna) [reply pynxtools/dataconverter/readers/ellips/reader.py:308]: Yes, use [`math.isnan`](https://docs.python.org/3/library/math.html#math.isnan)

File: pynxtools/dataconverter/readers/ellips/reader.py (hunk 2 above)
Comment (domna) [reply pynxtools/dataconverter/readers/ellips/reader.py:95]: Ok, if the empty list produces an error we should fix this in the `flatten_and_replace` function as it should also be possible to leave this empty.

I just looked at the code there and it just uses the `items()` method on this replace nested dict, which should be fine for an empty list. So I don't understand where the error was coming from. Could you post your error message you get?

File: pynxtools/dataconverter/readers/ellips/reader.py [hunk 7]
```diff
@@ -43,21 +83,16 @@ def load_header(filename, default):
...
-            clean_header[key] = val
+    clean_header = header
 
     for key, value in default.items():
         if key not in clean_header:
             clean_header[key] = value
+
+    for key, val in header.items():
+        if key == 'sep':
+            clean_header[key] = val.encode("utf-8").decode("unicode_escape")
```
Comment (domna) [inline pynxtools/dataconverter/readers/ellips/reader.py:94]:
```suggestion
    if 'sep' in header:
        clean_header['sep'] = header['sep'].encode("utf-8").decode("unicode_escape")
```

**Review by domna — APPROVED** [review]

**Comment by domna** [comment]
Please also remember to update the `eln_data.yaml` and eln files in the north repository (https://gitlab.mpcdf.mpg.de/nomad-lab/nomad-remote-tools-hub/-/tree/develop/docker/ellips/example)

File: pynxtools/dataconverter/readers/ellips/reader.py (hunk 3 above)
Comment (domna) [inline pynxtools/dataconverter/readers/ellips/reader.py]:
```suggestion
    while math.isnan(temp):
```

Comment (cmmngr) [reply pynxtools/dataconverter/readers/ellips/reader.py]: already done

Comment (domna) [reply pynxtools/dataconverter/readers/ellips/reader.py]: Ah sorry. The diff was alread showing this. But gh diff sometimes gets stuck if there are merge conflicts... (or maybe it's a feature I'm to stupid to understand 🤔)

File: examples/ellipsometry/eln_data.yaml [hunk 8]
```diff
@@ -1,71 +1,73 @@
-acquisition_program/\@url: https://www.jawoollam.com/ellipsometry-software/wvase
```
Comment (domna) [inline examples/ellipsometry/eln_data.yaml:1]: I would suggest to use an eln_data.yaml here which was generated by nomad

File: pynxtools/dataconverter/readers/ellips/reader.py [hunk 9]
```diff
@@ -18,16 +18,56 @@
...
+    'Instrument': 'INSTRUMENT[instrument]',
+    'Sample': 'SAMPLE[sample]',
+    'Stage': 'sample_stage',
+    'User': 'USER[user]',
+    'Instrument/angle_of_incidence': 'INSTRUMENT[instrument]/angle_of_incidence',
+    'Instrument/angle_of_incidence/unit': 'INSTRUMENT[instrument]/angle_of_incidence/@units',
+    'column_names': 'data_collection/column_names',
+    'data_error': 'data_collection/data_error',
+    'depolarization': 'derived_parameters/depolarization',
+    'measured_data': 'data_collection/measured_data',
```
Comment (domna) [inline pynxtools/dataconverter/readers/ellips/reader.py:50]:
```suggestion
    'measured_data': 'data_collection/measured_data',
    'software': 'software/program',
    'data_software': 'data_software/program',
```

**Pull request merged**